File: WeatherSense/trunk/VantagePro2Driver/src/vantagepro2/ParametersMessage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;

namespace VantagePro2 {
    /// <summary>
    /// Message to send weather station parameters to the collector
    /// </summary>
    public class ParametersMessage {
        private readonly static String MANUFACTURER = "Davis Instruments";
        private readonly static String MODEL = "Vantage Pro 2";
        private readonly static String KEY_VALUE = "<entry><key>{0}</key><value>{1}</value></entry>";

        private String firmwareDate;
        public String FirmwareDate {
            get { return firmwareDate; }
            set { firmwareDate = value; }
        }

        private String firmwareVersion;
        public String FirmwareVersion {
            get { return firmwareVersion; }
            set { firmwareVersion = value; }
        }

        private double latitude;
        public double Latitude {
            get { return latitude; }
            set { latitude = value; }
        }

        private double longitude;
        public double Longitude {
            get { return longitude; }
            set { longitude = value; }
        }

        private int elevation;
        public int Elevation {
            get { return elevation; }
            set { elevation = value; }
        }

        private int rainSeasonStart;
        public int RainSeasonStart {
            get { return rainSeasonStart; }
            set { rainSeasonStart = value; }
        }

        private int archivePeriod;
        public int ArchivePeriod {
            get { return archivePeriod; }
            set { archivePeriod = value; }
        }

        private int windCupSize;
        public int WindCupSize {
            get { return windCupSize; }
            set { windCupSize = value; }
        }

        private float rainCollectorSize;
        public float RainCollectorSize {
            get { return rainCollectorSize; }
            set { rainCollectorSize = value; }
        }

 
        /// <summary>
        /// Format the parameters message
        /// </summary>
        /// <returns>The message as a string</returns>
        public String formatMessage() {
            StringBuilder sb = new StringBuilder("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>");
            sb.Append("<wsParametersMessage>");
            sb.AppendFormat("<manufacturer>{0}</manufacturer>", MANUFACTURER);
            sb.AppendFormat("<model>{0}</model>", MODEL);
            sb.AppendFormat("<firmwareDate>{0}</firmwareDate>", firmwareDate);
            sb.AppendFormat("<firmwareVersion>{0}</firmwareVersion>", firmwareVersion);
            sb.AppendFormat("<location><latitude><value>{0}</value></latitude>", latitude);
            sb.AppendFormat("<longitude><value>{0}</value></longitude>", longitude);
            sb.AppendFormat("<altitude><value>{0}</value></altitude></location>", UnitConverter.feetToMillimeters(elevation));
            sb.AppendFormat("<weatherYearStartMonth>{0}</weatherYearStartMonth>", RainSeasonStart);
            sb.Append("<parameters>");
            sb.AppendFormat(KEY_VALUE, "Archive Period", archivePeriod);
            sb.AppendFormat(KEY_VALUE, "Wind Cup Size", windCupSize == 0 ? "Small" : "Large");
            sb.AppendFormat(KEY_VALUE, "Rain Collector Size", rainCollectorSize + " inches");
            sb.Append("</parameters>");
            sb.Append("</wsParametersMessage>");

            return sb.ToString();
        }
    }
}

File: VantagePro2Driver/VP2++/src/VP2/ParametersMessage.cpp
/* 
 * Copyright (C) 2020 Bruce Beisel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <sstream>
#include "UnitConverter.h"
#include "ParametersMessage.h"

using namespace std;

namespace vp2 {
const std::string ParametersMessage::MANUFACTURER = "Davis Instruments";
const std::string ParametersMessage::MODEL = "Vantage Pro 2";

ParametersMessage::ParametersMessage() {
}

ParametersMessage::~ParametersMessage() {
}

string
ParametersMessage::getFirmwareDate() const {
    return firmwareDate;
}

void
ParametersMessage::setFirmwareDate(const string & value) {
    firmwareDate = value;
}

string
ParametersMessage::getFirmwareVersion() const {
    return firmwareVersion;
}

void
ParametersMessage::setFirmwareVersion(const string & value) {
    firmwareVersion = value;
}

double
ParametersMessage::getLatitude() const {
    return latitude;
}

void
ParametersMessage::setLatitude(double value) {
    latitude = value;
}

double
ParametersMessage::getLongitude() const {
    return longitude;
}

void
ParametersMessage::setLongitude(double value) {
    longitude = value;
}

int
ParametersMessage::getElevation() const {
    return elevation;
}

void
ParametersMessage::setElevation(int value) {
    elevation = value;
}

int
ParametersMessage::getRainSeasonStart() const {
    return rainSeasonStart;
}

void
ParametersMessage::setRainSeasonStart(int value) {
    rainSeasonStart = value;
}

int
ParametersMessage::getArchivePeriod() const {
    return archivePeriod;
}

void
ParametersMessage::setArchivePeriod(int value) {
    archivePeriod = value;
}

int
ParametersMessage::getWindCupSize() const {
    return windCupSize;
}

void
ParametersMessage::setWindCupSize(int value) {
    windCupSize = value;
}

Rainfall
ParametersMessage::getRainCollectorSize() const {
    return rainCollectorSize;
}

void
ParametersMessage::setRainCollectorSize(Rainfall value) {
    rainCollectorSize = value;
}

/// <summary>
/// Format the parameters message
/// </summary>
/// <returns>The message as a string</returns>
string
ParametersMessage::formatMessage() const {
    ostringstream ss;
    ss << "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>";
    ss << "<wsParametersMessage>";
    ss << "<manufacturer>" << MANUFACTURER << "</manufacturer>";
    ss << "<model>" << MODEL << "</model>";
    ss << "<firmwareDate>" << firmwareDate << "</firmwareDate>";
    ss << "<firmwareVersion>" << firmwareVersion << "</firmwareVersion>";
    ss << "<location><latitude>" << latitude << "</latitude>";
    ss << "<longitude>" << longitude << "</longitude>";
    ss << "<altitude>" << UnitConverter::feetToMillimeters(elevation) << "</altitude></location>";
    ss << "<weatherYearStartMonth>" << rainSeasonStart << "</weatherYearStartMonth>";
    ss << "<parameters>";
    ss << "<entry><key>" << "Archive Period" << "</key><value>" << archivePeriod << "</value></entry>";
    ss << "<entry><key>" << "Wind Cup Size" << "</key><value>" << (windCupSize == 0 ? "Small" : "Large") << "</value></entry>";
    ss << "<entry><key>" << "Rain Collector Size" << "</key><value>" << rainCollectorSize << " inches" << "</value></entry>";
    ss << "</parameters>";
    ss << "</wsParametersMessage>";

    return ss.str();
}
}
